clazy: fix darwin builds

marcusramberg opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Allow clazy to be built on Darwin by patching a brew specific check.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

marcusramberg committed:
clazy: fix darwin builds

stigtsp commented on pkgs/development/tools/analysis/clazy/default.nix:
```diff
@@ -28,6 +28,11 @@ stdenv.mkDerivation rec {
 
+  postPatch = ''
+    # clazy expects brew readlink on macOS, but we have the GNU version
+    substituteInPlace CMakeLists.txt --replace-fail greadlink readlink
```
Maybe this should refer to `coreutils` as well

```suggestion
    substituteInPlace CMakeLists.txt --replace-fail greadlink ${coreutils}/bin/readlink
```

ofalvai commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366479`

---
### `x86_64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>clazy</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>clazy</li>
  </ul>
</details>

ofalvai commented:
Both platforms fail for me with:
```
clazy-x86_64-darwin> Running phase: buildPhase
clazy-x86_64-darwin> build flags: -j14 SHELL=/nix/store/45a4hms3bsmk23bzn0g0qfn663izq3f8-bash-5.2p37/bin/bash
clazy-x86_64-darwin> [  0%] Building CXX object CMakeFiles/ClazyPlugin.dir/cmake_pch.hxx.pch
clazy-x86_64-darwin> In file included from <built-in>:1:
clazy-x86_64-darwin> In file included from /tmp/nix-build-clazy-1.12.drv-1/source/build/CMakeFiles/ClazyPlugin.dir/cmake_pch.hxx:5:
clazy-x86_64-darwin> /tmp/nix-build-clazy-1.12.drv-1/source/src/checkbase.h:99:73: error: non-virtual member function marked 'override' hides virtual member function
clazy-x86_64-darwin>    99 |                             clang::SrcMgr::CharacteristicKind FileType) override;
clazy-x86_64-darwin>       |                                                                         ^
clazy-x86_64-darwin> /nix/store/mhqm54z1jkwaxj1kcfvqlyaxkdljwiwl-clang-19.1.5-dev/include/clang/Lex/PPCallbacks.h:168:16: note: hidden overloaded virtual function 'clang::PPCallbacks::InclusionDirective' declared here: different number of parameters (11 vs 10)
clazy-x86_64-darwin>   168 |   virtual void InclusionDirective(SourceLocation HashLoc,
clazy-x86_64-darwin>       |                ^
clazy-x86_64-darwin> 1 error generated.
clazy-x86_64-darwin> make[2]: *** [CMakeFiles/ClazyPlugin.dir/build.make:77: CMakeFiles/ClazyPlugin.dir/cmake_pch.hxx.pch] Error 1
clazy-x86_64-darwin> make[1]: *** [CMakeFiles/Makefile2:117: CMakeFiles/ClazyPlugin.dir/all] Error 2
clazy-x86_64-darwin> make: *** [Makefile:166: all] Error 2
```

You might want to rebase your branch onto the latest master because the Clang version was bumped in the latest staging-next iteration, and I think this error is related.

marcusramberg commented:
> Both platforms fail for me with:
> 
> ```
> clazy-x86_64-darwin> Running phase: buildPhase
> clazy-x86_64-darwin> build flags: -j14 SHELL=/nix/store/45a4hms3bsmk23bzn0g0qfn663izq3f8-bash-5.2p37/bin/bash
> clazy-x86_64-darwin> [  0%] Building CXX object CMakeFiles/ClazyPlugin.dir/cmake_pch.hxx.pch
> clazy-x86_64-darwin> In file included from <built-in>:1:
> clazy-x86_64-darwin> In file included from /tmp/nix-build-clazy-1.12.drv-1/source/build/CMakeFiles/ClazyPlugin.dir/cmake_pch.hxx:5:
> clazy-x86_64-darwin> /tmp/nix-build-clazy-1.12.drv-1/source/src/checkbase.h:99:73: error: non-virtual member function marked 'override' hides virtual member function
> clazy-x86_64-darwin>    99 |                             clang::SrcMgr::CharacteristicKind FileType) override;
> clazy-x86_64-darwin>       |                                                                         ^
> clazy-x86_64-darwin> /nix/store/mhqm54z1jkwaxj1kcfvqlyaxkdljwiwl-clang-19.1.5-dev/include/clang/Lex/PPCallbacks.h:168:16: note: hidden overloaded virtual function 'clang::PPCallbacks::InclusionDirective' declared here: different number of parameters (11 vs 10)
> clazy-x86_64-darwin>   168 |   virtual void InclusionDirective(SourceLocation HashLoc,
> clazy-x86_64-darwin>       |                ^
> clazy-x86_64-darwin> 1 error generated.
> clazy-x86_64-darwin> make[2]: *** [CMakeFiles/ClazyPlugin.dir/build.make:77: CMakeFiles/ClazyPlugin.dir/cmake_pch.hxx.pch] Error 1
> clazy-x86_64-darwin> make[1]: *** [CMakeFiles/Makefile2:117: CMakeFiles/ClazyPlugin.dir/all] Error 2
> clazy-x86_64-darwin> make: *** [Makefile:166: all] Error 2
> ```
> 
> You might want to rebase your branch onto the latest master because the Clang version was bumped in the latest staging-next iteration, and I think this error is related.

Didn't see those errors after rebasing against current master with recent clang 

(`-- Found Clang (LLVM version: 19.1.5)`) - however, I noticed the checkphase was failing due to lack of awk/dirname/head, so I added toybox and gawk to the path, seems to build fine here on arm64 darwin.

marcusramberg commented:
```
┏━ Dependency Graph:
┃ ┌─ ↓ ✔ stdenv-darwin
┃ ├─ ✔ clazy-1.13 ⏱ 25s
┃ ✔ review-shell
┣━━━ Builds          │ Downloads       │ Host
┃        │ ✔ 2 │     │     │     │     │ localhost
┃        │     │     │     │ ↓ 1 │     │ https://cache.nixos.org
┗━ ∑ ⏵ 0 │ ✔ 2 │ ⏸ 0 │ ↓ 0 │ ↓ 1 │ ⏸ 0 │ Finished at 20:22:38 after 29s
--------- Report for 'aarch64-darwin' ---------
1 package built:
clazy
```

ofalvai commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366479`

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>clazy</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>clazy</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>clazy</li>
  </ul>
</details>

marcusramberg commented:
> > however, I noticed the checkphase was failing due to lack of awk/dirname/head, so I added toybox and gawk to the path, seems to build fine here on arm64 darwin.
> 
> If these are only needed during `checkPhase`, couldn't they go into `nativeCheckInputs` instead? I'm not worried too much about closure size because it's dominated by Clang anyway, but it seems like a quick and simple change.

I tested and I couldn't get that to work (still didn't find the commands), but I replaced toybox with coreutils.
